src/model/model.ts

```diff
@@ -1,43 +1,105 @@
-import { getCollection, validateSchema } from '..';
-import { save } from '../handler/save';
-import { find } from '../handler/find';
 import { generateUUID } from '../utils/generate-uuid';
+import { extractDataFromModel } from '../utils/extract-data-from-model';
+import { storeLifeCycle } from './store-life-cycle';
+import { COLLECTION_KEY } from '../utils/constants';
+import { removeLifeCicle } from './remove-life-cycle';
 
+/**
+ * Constructor to build a model instance base on schema and options
+ * Provide methods to handle documents of the current collection on DB
+ */
 export class Model {
-  __schema;
-  __collection;
-  data;
+  /**
+   * @ignore
+   */
+  $!: {
+    collectionName: string;
+    schema: any;
+    collection: any;
+    ID_KEY: string;
+  };
 
   constructor(data: any) {
-    this.data = data;
+    for (const key in data) {
+      this[key] = data[key];
+    }
   }
 
-  save() {
-    validateSchema(this.data, this.__schema);
-    return save(generateUUID(), this.data, this.__collection);
+  /**
+   * return id value
+   */
+  getId(): string {
+    return this[this.$.ID_KEY];
   }
 
-  static find = (params): Promise<any> => Promise.resolve();
-  static create = (data: any): Promise<any> => Promise.resolve();
-}
-
-export const model = (name, schema?, connection?, options?) => {
-  const collection = getCollection(name);
-  return class ModelFactory extends Model {
-    constructor(data) {
-      super(data);
-      this.__collection = collection;
-      this.__schema = schema;
+  /**
+   * Save or Update document, I
+   * If the document has id (key) will be update
+   * Else a new id (key) will be assigned and document will be stored
+   */
+  async save() {
+    const data = extractDataFromModel(this);
+    const options: any = {};
+    const id = this.getId();
+    if (!id) {
+      data[this.$.ID_KEY] = generateUUID();
+    } else {
+      const { cas } = await this.$.collection.get(id);
+      if (cas) {
+        options.cas = cas;
+      }
     }
+    const addedMetadata = { ...data, [COLLECTION_KEY]: this.$.collectionName };
+    return storeLifeCycle({ data: addedMetadata, options, model: this.$ });
+  }
 
-    static find = (params): Promise<any> => {
-      return find(params, collection);
-    };
+  /**
+   * Remove document from BD
+   */
+  remove(options = {}) {
+    return removeLifeCicle({ id: this.getId(), options, model: this.$ });
+  }
 
-    static create = (data: Record<string, any>): Promise<any> => save(generateUUID(), data, collection);
+  /**
+   * Return javascript object with data
+   */
+  toCoo() {
+    return extractDataFromModel(this);
+  }
 
-    static insertMany(data: Record<string, any>[]): Promise<any> {
-      return Promise.all(data.map((d) => this.create(d)));
-    }
-  };
-};
+  /**
+   * Return javascript object serialized to JSON
+   */
+  toJSON() {
+    return JSON.stringify(this.toCoo());
+  }
+
+  /**
+   * Find documents base on filter on the collection
```

Comment:
Should be: `Find documents (based) on filter (of) the collection`

Reply:
fixed the typos